src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_azgalor.cpp

```diff
@@ -18,269 +18,115 @@
 #include "CreatureScript.h"
 #include "ScriptedCreature.h"
 #include "hyjal.h"
-#include "hyjal_trash.h"
 
 enum Spells
 {
     SPELL_RAIN_OF_FIRE          = 31340,
     SPELL_DOOM                  = 31347,
     SPELL_HOWL_OF_AZGALOR       = 31344,
     SPELL_CLEAVE                = 31345,
-    SPELL_BERSERK               = 26662,
-
-    SPELL_THRASH                = 12787,
-    SPELL_CRIPPLE               = 31406,
-    SPELL_WARSTOMP              = 31408,
+    SPELL_BERSERK               = 26662
 };
 
 enum Texts
 {
     SAY_ONDEATH             = 0,
     SAY_ONSLAY              = 1,
     SAY_DOOM                = 2, // Not used?
-    SAY_ONAGGRO             = 3,
+    SAY_ONSPAWN             = 3,
+
+    SAY_ARCHIMONDE_INTRO    = 8
 };
 
-class boss_azgalor : public CreatureScript
+struct boss_azgalor : public BossAI
 {
 public:
-    boss_azgalor() : CreatureScript("boss_azgalor") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
+    boss_azgalor(Creature* creature) : BossAI(creature, DATA_AZGALOR)
     {
-        return GetHyjalAI<boss_azgalorAI>(creature);
+        _recentlySpoken = false;
+        scheduler.SetValidator([this]
+            {
+                return !me->HasUnitState(UNIT_STATE_CASTING);
+            });
     }
 
-    struct boss_azgalorAI : public hyjal_trashAI
+    void JustEngagedWith(Unit * who) override
     {
-        boss_azgalorAI(Creature* creature) : hyjal_trashAI(creature)
-        {
-            instance = creature->GetInstanceScript();
-            go = false;
-        }
-
-        uint32 RainTimer;
-        uint32 DoomTimer;
-        uint32 HowlTimer;
-        uint32 CleaveTimer;
-        uint32 EnrageTimer;
-        bool enraged;
-
-        bool go;
+        BossAI::JustEngagedWith(who);
 
-        void Reset() override
+        scheduler.Schedule(10s, 16s, [this](TaskContext context)
```

Comment:
btw you can also use ScheduleTimedEvent and ScheduleUniqueTimedEvent for scheduling abilities now. But it's not a must

Reply:
Interesting... I didn't know about those. I'll try to look into if I get further into Black Temple